### PR TITLE
fix crash while saving attachments in MongoTrials

### LINKED ISSUE #624: Error: 'Attachments' object has no attribute 'handle'
I get a error while using `MongoTrails` (this works fine with `Trails`).

Using `hyperopt==0.2.3` with Python 3.6.5

```
import pickle
import time
from hyperopt import fmin, tpe, hp, STATUS_OK, Trials

def objective(x):
    return {
        'loss': x ** 2,
        'status': STATUS_OK,
        # -- store other results like this
        'eval_time': time.time(),
        'other_stuff': {'type': None, 'value': [0, 1, 2]},
        # -- attachments are handled differently
        'attachments':
            {'time_module': pickle.dumps(time.time)}
        }
trials = MongoTrials("mongo://localhost:27017/foo_db3/jobs", exp_key="exp1")
best = fmin(objective,
    space=hp.uniform('x', -10, 10),
    algo=tpe.suggest,
    max_evals=100,
    trials=trials)

print(best)
```

Error from the worker:

```
Traceback (most recent call last):
  File "/Users/sam/.pyenv/versions/d4r/bin/hyperopt-mongo-worker", line 8, in <module>
    sys.exit(main())
  File "/Users/sam/.pyenv/versions/3.6.5/envs/d4r/lib/python3.6/site-packages/hyperopt/mongoexp.py", line 1331, in main
    sys.exit(main_worker())
  File "/Users/sam/.pyenv/versions/3.6.5/envs/d4r/lib/python3.6/site-packages/hyperopt/mongoexp.py", line 1413, in main_worker
    return main_worker_helper(options, args)
  File "/Users/sam/.pyenv/versions/3.6.5/envs/d4r/lib/python3.6/site-packages/hyperopt/mongoexp.py", line 1324, in main_worker_helper
    mworker.run_one(reserve_timeout=float(options.reserve_timeout))
  File "/Users/sam/.pyenv/versions/3.6.5/envs/d4r/lib/python3.6/site-packages/hyperopt/mongoexp.py", line 1112, in run_one
    result = worker_fn(spec, ctrl)
  File "/Users/sam/.pyenv/versions/3.6.5/envs/d4r/lib/python3.6/site-packages/hyperopt/base.py", line 914, in evaluate
    ctrl.attachments[key] = val
  File "/Users/sam/.pyenv/versions/3.6.5/envs/d4r/lib/python3.6/site-packages/hyperopt/mongoexp.py", line 943, in __setitem__
    self.handle.set_attachment(
AttributeError: 'Attachments' object has no attribute 'handle'
```

### PR DESCRIPTION
**Pull request by codereverser**
fixes #624

The "Trials object" example @ https://hyperopt.github.io/hyperopt/getting-started/minimizing_functions/#the-trials-object fails with MongoTrials since Attachments has no access to MongoTrials.handle

fixes the issue (#624) and also updates a test case to demonstrate the issue.

### CODE REVIEW

File: hyperopt/tests/test_mongoexp.py [hunk 1]
```diff
@@ -256,6 +257,48 @@ def test_attachments(trials):
...
+    def fmin_thread_fn(space, mongo_trials, evals):
+        fmin(
+            fn=passthrough_with_attachments,
+            space=space,
+            algo=rand.suggest,
+            trials=mongo_trials,
+            rstate=np.random.RandomState(),
+            max_evals=evals,
+            return_argmin=False,
+        )
```
Comment (marctorsoc) [inline hyperopt/tests/test_mongoexp.py:270]: does this need to be nested function? can't it be declared after the test?

File: hyperopt/tests/test_mongoexp.py [hunk 2]
```diff
@@ -286,6 +329,13 @@ def passthrough(x):
     return x
 
 
+def passthrough_with_attachments(x):
```
Comment (marctorsoc) [inline hyperopt/tests/test_mongoexp.py:332]: could you give a better name than x and typehint it? Also could you move this function after `test_trial_attachments` ?

**Review by marctorsoc — APPROVED** [review]
Thanks for your PR! Just some minor comments :)

**Comment by codereverser** [comment]
Thanks for the review. I've added the type-hints and refactored the new test functions a bit

**Pull request merged**